gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java

```diff
@@ -0,0 +1,282 @@
+import static jakarta.ws.rs.core.MediaType.TEXT_PLAIN_TYPE;
+import static jakarta.ws.rs.core.Response.Status.BAD_GATEWAY;
+import static jakarta.ws.rs.core.Response.Status.OK;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.list;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
+public class ProxyRequestHandler
+{
+    private static final Logger log = Logger.get(ProxyRequestHandler.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private final Duration asyncTimeout;
+    private final ExecutorService executor = newCachedThreadPool(daemonThreadsNamed("proxy-%s"));
+    private final HttpClient httpClient;
+    private final RoutingManager routingManager;
+    private final QueryHistoryManager queryHistoryManager;
+    private final boolean cookiesEnabled;
+
+    @Inject
+    public ProxyRequestHandler(
+            @ForProxy HttpClient httpClient,
+            RoutingManager routingManager,
+            QueryHistoryManager queryHistoryManager,
+            HaGatewayConfiguration haGatewayConfiguration)
+    {
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.routingManager = requireNonNull(routingManager, "routingManager is null");
+        this.queryHistoryManager = requireNonNull(queryHistoryManager, "queryHistoryManager is null");
+        cookiesEnabled = GatewayCookieConfigurationPropertiesProvider.getInstance().isEnabled();
+        asyncTimeout = haGatewayConfiguration.getRouting().getAsyncTimeout();
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        executor.shutdownNow();
+    }
+
+    public void getRequest(
+            HttpServletRequest servletRequest,
+            AsyncResponse asyncResponse,
+            URI remoteUri)
+    {
+        Request.Builder request = prepareGet();
+        performRequest(remoteUri, servletRequest, asyncResponse, request);
+    }
+
+    public void postRequest(
+            String statement,
+            HttpServletRequest servletRequest,
+            AsyncResponse asyncResponse,
+            URI remoteUri)
+    {
+        Request.Builder request = preparePost()
+                .setBodyGenerator(createStaticBodyGenerator(statement, UTF_8));
+        performRequest(remoteUri, servletRequest, asyncResponse, request);
+    }
+
+    private void performRequest(
+            URI remoteUri,
+            HttpServletRequest servletRequest,
+            AsyncResponse asyncResponse,
+            Request.Builder requestBuilder)
+    {
+        requestBuilder.setUri(remoteUri);
+
+        for (String name : list(servletRequest.getHeaderNames())) {
+            for (String value : list(servletRequest.getHeaders(name))) {
+                // TODO: decide what should and shouldn't be forwarded
```

Comment:
Does this need to be done in this PR? Or is this done already and we can remove?

Reply:
That depends on the actual deployment setup.
Different user / vendor might use headers to do different kind of stuffs, like security/routing/tracing/etc.
I'd like to address this in other PR with more discussion.

Reply:
Sounds good.